Annotate admin_order_field to also accept a BaseExpression

brianhelba opened the pull request:
[Django's docs state](https://docs.djangoproject.com/en/dev/ref/contrib/admin/#django.contrib.admin.ModelAdmin.list_display):
> Query expressions may be used in `admin_order_field`.

Internally, [Django looks for the property `resolve_expression`](https://github.com/django/django/blob/509d9da26fb92a8d566ec105ff40bb024803ceaa/django/contrib/admin/views/main.py#L319-L326) on non-string `admin_order_field` values. This [is provided by `BaseExpression`](https://github.com/django/django/blob/master/django/db/models/expressions.py#L231).

escaped reviewed (changes requested):
Nice find @brianhelba! Thank you. 

Implementation-wise it looks great, but I think the new functionality should be covered by the tests as well. Would you mind integrating the new case into the tests? If you need any help please let me know.

(...and while you at it, don't forget to add yourself to the `CONTRIBUTORS.md` :) )

brianhelba commented:
@escaped Thanks for the quick review and feedback.
* I've added myself to `CONTRIBUTORS.md`
* I've added a sample usage to the tests, using the case from [the Django docs](https://docs.djangoproject.com/en/3.1/ref/contrib/admin/#django.contrib.admin.ModelAdmin.list_display)

brianhelba commented:
@escaped The tests are failing, but it looks to me like it might be an issue with `master`, not this branch. I'm not sure though.

escaped commented on tests/test_decorator.py:
```diff
@@ -1,4 +1,6 @@
 import django
+from django.db.models import Value
+from django.db.models.functions import Concat
```
![image](https://user-images.githubusercontent.com/318650/97602140-33e91900-19c8-11eb-810c-0e13aad1faac.png)

Looks like the order is not correct and `isort` would like to fix this as it is configured in a way that `import ...` statements have to be before `from ... import ...`.

There is a `pre-commit` hook in place, which you can use to automatically fix the file. It should run automatically on every commit after it had been installed using `poetry run pre-commit install`. However, you can also run it manually by running `poetry run pre-commit run -a` and it should automatically change the order of the import to the expected format.

escaped commented on tests/test_decorator.py:
```diff
@@ -9,6 +11,7 @@
 
 OPTIONS = [
     ('admin_order_field', 'radius'),
+    ('admin_order_field', Concat('first_name', Value(' '), 'last_name')),
```
👍  Tricky bit is, that we test mypy using some python code which is encoded as a string:

- https://github.com/escaped/django-admin-display/blob/52dbef44dc5804c45e1d748088e32568fddb9c90/tests/test_mypy.py#L10
- https://github.com/escaped/django-admin-display/blob/52dbef44dc5804c45e1d748088e32568fddb9c90/tests/test_mypy.py#L29

As we use `OPTIONS` in the corresponding parametrize, we need to add the imports for `Value` and `Concat` there as well. This should fix the error stated in the logs

![image](https://user-images.githubusercontent.com/318650/97603020-3dbf4c00-19c9-11eb-9161-18a3d7905111.png)

escaped reviewed (changes requested):
Nice work. Some minor changes and the PR is ready to be merged. 

I inspected the output of the CI pipeline and it looks like we need to make some minor adjustments to make it pass.
Furthermore, yes, there is an issue with `master` which states that `isort` is not configured correctly. I have to fix this in another project, but as it is just a warning and it should not affect this PR. 

Please have a look at my comments and let me know if you need any help. Thank you.

brianhelba commented:
@escaped Thanks for all the guidance.

The good news is that I found a way to further simplify the tests. [Using `repr`](https://docs.python.org/3/library/functions.html#repr) for injected values is simpler than the special case of having to explicitly wrap string values in quotes.

The bad news is that Django's use of `repr` for query expressions is broken, as the returned value does not properly quote parameters. So, I had to add another special case to fix this.

escaped commented on tests/test_mypy.py:
```diff
@@ -1,21 +1,44 @@
 from .test_decorator import OPTIONS
 
 
+def safe_repr(value):
```
Nice work! Today I learned something new :)

I like the solution; it's clever, but maybe too clever in a way that it will be hard for others to understand what's going on. So, I am a bit on a fence, and maybe we should just be explicit and redefine everything instead of importing the `OPTIONS`.

```python
OPTIONS = [
    ('admin_order_field', '"radius"'),
    ('admin_order_field', 'Lower(F("person_name"))'),
    ('boolean', 'True'),
    ('empty_value_display', '"Undefined"'),
    ('short_description', '"Is big?"'),
]
if django.VERSION[:2] <= (1, 11):
    OPTIONS.append(('allow_tags', "True"))
```

I know you put a lot of effort into solving this issue, so I am curious to hear what you think about this.

escaped reviewed (approved):
Great work. Thank you, it's ready for merging :)  Will definitely publish it this week (even though I left one discussion point).
